### ISSUE TITLE
Optimize shifts better

### ISSUE THREAD

**Issue by GSPP** [dotnet/coreclr#1741]
Let's look at the following code performing some shifts:

        [MethodImpl(MethodImplOptions.NoInlining)]
        static void Shift()
        {
            var x = Havoc(0x1234);
            var y = Havoc(0x1235);

            Sink(x << y);
            Sink(x << (y & 0x1F));
            Sink(x << (y & 0x1));
        }
        [MethodImpl(MethodImplOptions.NoInlining)]
        static int Havoc(int x)
        {
            return x;
        }
        [MethodImpl(MethodImplOptions.NoInlining)]
        static void Sink(int x)
        {
        }

Under VS15, x64, Release this compiles to:

```
            var x = Havoc(0x1234);
00000000  push        rdi 
00000001  push        rsi 
00000002  push        rbx 
00000003  sub         rsp,20h 
00000007  mov         ecx,1234h 
0000000c  call        FFFFFFFFFFFFECD0 //havoc
00000011  mov         esi,eax 
00000013  mov         ecx,1235h 
00000018  call        FFFFFFFFFFFFECD0 //havoc
0000001d  mov         edi,eax 
0000001f  mov         ebx,edi //unnecessary register move
00000021  and         ebx,1Fh
00000024  mov         ecx,ebx 
00000026  mov         eax,esi  //unnecessary register move
00000028  shl         eax,cl 
0000002a  mov         ecx,eax 
0000002c  call        FFFFFFFFFFFFEAD0 
            Sink(x << (y & 0x1F));
00000031  mov         ecx,ebx 
00000033  and         ecx,1Fh 
00000036  mov         eax,esi  //unnecessary register move
00000038  shl         eax,cl 
0000003a  mov         ecx,eax  //unnecessary register move
0000003c  call        FFFFFFFFFFFFEAD0 
            Sink(x << (y & 0x1));
00000041  mov         ecx,edi 
00000043  and         ecx,1 
00000046  and         ecx,1Fh //should be combined with the and above
00000049  mov         eax,esi  //unnecessary register move
0000004b  shl         eax,cl 
0000004d  mov         ecx,eax  //unnecessary register move
0000004f  call        FFFFFFFFFFFFEAD0 
00000054  nop 
        }
00000055  add         rsp,20h 
00000059  pop         rbx 
0000005a  pop         rsi 
0000005b  pop         rdi 
0000005c  ret 
```

I have marked some code gen issues. I think it's worth looking into those. The unnecessary moves might be an issue that is present in other code patterns as well.

**Comment by Janiels** [comment]
Do you have any benchmarks showing the gain of eliminating the moves?
The CPU can perform them via simple register renames and optimizing them away is probably just a code size optimization.

**Comment by mikedn** [comment]
Some additions/corrections:

```
            var x = Havoc(0x1234);
00000000  push        rdi 
00000001  push        rsi 
00000002  push        rbx 
00000003  sub         rsp,20h 
00000007  mov         ecx,1234h 
0000000c  call        FFFFFFFFFFFFECD0 //havoc
00000011  mov         esi,eax 
00000013  mov         ecx,1235h 
00000018  call        FFFFFFFFFFFFECD0 //havoc
0000001d  mov         edi,eax 
// true but only if various other issues are fixed first
// as is now this copy is necessary because there edi is used later
0000001f  mov         ebx,edi //unnecessary register move 
00000021  and         ebx,1Fh
00000024  mov         ecx,ebx 
// this copy is necessary, esi is used later so it cannot be modified here
00000026  mov         eax,esi  //unnecessary register move
00000028  shl         eax,cl 
0000002a  mov         ecx,eax 
0000002c  call        FFFFFFFFFFFFEAD0 
            Sink(x << (y & 0x1F));
00000031  mov         ecx,ebx 
00000033  and         ecx,1Fh 
// this copy is necessary, esi is used later so it cannot be modified here
00000036  mov         eax,esi  //unnecessary register move
00000038  shl         eax,cl 
// this copy is necessary, the first function argument is passed in ecx
0000003a  mov         ecx,eax  //unnecessary register move
0000003c  call        FFFFFFFFFFFFEAD0 
            Sink(x << (y & 0x1));
00000041  mov         ecx,edi 
00000043  and         ecx,1 
// sort of, it really needs to be eliminated from here and all previous cases
// the shl instruction masks the shift count anyway
00000046  and         ecx,1Fh //should be combined with the and above
// this is indeed unnecessary, esi is no longer used
00000049  mov         eax,esi  //unnecessary register move
0000004b  shl         eax,cl 
// this copy is necessary, the first function argument is passed in ecx
0000004d  mov         ecx,eax  //unnecessary register move
0000004f  call        FFFFFFFFFFFFEAD0 
00000054  nop 
        }
00000055  add         rsp,20h 
00000059  pop         rbx 
0000005a  pop         rsi 
0000005b  pop         rdi 
0000005c  ret 
```

The main problem in all this of code is really `and ecx, 1Fh`. It's useless because `shl` masks the shift count anyway.

**Comment by mikedn** [comment]
Here's the same code compiled with a patched JIT:

```
       B934120000           mov      ecx, 0x1234
       E8BFFBFFFF           call     Program:Havoc(int):int
       8BF0                 mov      esi, eax
       B935120000           mov      ecx, 0x1235
       E8B3FBFFFF           call     Program:Havoc(int):int
       8BF8                 mov      edi, eax
       8BCF                 mov      ecx, edi
       8BDE                 mov      ebx, esi
       D3E3                 shl      ebx, cl
       8BCB                 mov      ecx, ebx
       E8ACFBFFFF           call     Program:Sink(int)
       8BCB                 mov      ecx, ebx
       E8A5FBFFFF           call     Program:Sink(int)
       8BCF                 mov      ecx, edi
       83E101               and      ecx, 1
       D3E6                 shl      esi, cl
       8BCE                 mov      ecx, esi
       E897FBFFFF           call     Program:Sink(int)
```
Unless I'm missing something this is as good as it gets.

**Comment by mikedn** [comment]
> Unless I'm missing something this is as good as it gets.

Actually I'm missing something. `shlx` could be used instead of the lousy `shl` which requires the shift count to be placed in `cl`. It also avoids the need for a lot of moves due to its 3 operand form. :smile:

**Comment by GSPP** [comment]
@mikedn awesome analysis. I think this is suspicious:

       8BF8                 mov      edi, eax
       8BCF                 mov      ecx, edi

After these two instructions the value 0x1234 is now in three registers at the same time. How can that be necessary?! I think two are required.

       8BDE                 mov      ebx, esi
       D3E3                 shl      ebx, cl
       8BCB                 mov      ecx, ebx

Here, I think the shift should operate directly on ecx. That seems like a harder operation because it requires the JIT to understand that this value must efficiently arrive in the fixed ecx register used in the calling convention. The JIT now seems to use a strategy where it computes the value anywhere and is then forced to "patch it up" by copying to ecx.

**Comment by mikedn** [comment]
@GSPP The various moves are necessary because many registers have dedicated uses: `eax` is used to return values from functions, `ecx` is used to pass the first function argument, `esi`, `edi` and `ebx` are preserved across calls and the rest aren't, shift instructions require the shift count to be in `ecx` (`cl` represents the first 8 bits of `ecx`). With all these constraints in place there's little choice left:

```
mov      ecx, 0x1234
call     Program:Havoc(int):int
mov      esi, eax ; have to move, eax will get killed by the next call
mov      ecx, 0x1235
call     Program:Havoc(int):int
mov      edi, eax ; have to move, eax will get killed by calls but it's needed for the last shift
mov      ecx, edi ; have to move, shift requires the count in ecx
mov      ebx, esi ; have to move, esi is needed by the last shift
shl      ebx, cl ; also ebx is the ideal choice because it's call preserved
mov      ecx, ebx ; have to move, argument required in ecx
call     Program:Sink(int)
mov      ecx, ebx ; use the previously computed shift preserved in ebx
call     Program:Sink(int)
mov      ecx, edi ; have to move, shift requires count in ecx
and      ecx, 1
shl      esi, cl ; can compute the result in esi, no longer needed
mov      ecx, esi ; have to move, argument required in ecx
call     Program:Sink(int)
```

VC++ without AVX2 generates a very similar set of moves (though I had to help it a bit by removing & 0x1F, it too doesn't know that it isn't needed). What really helps is the `shlx` instruction which doesn't require the shift count in a particular register and avoids some of the other moves due to the 3 operand form. VC++ with AVX2 generates:

```
00C21023 B9 34 12 00 00       mov         ecx,1234h  
00C21028 E8 D3 FF FF FF       call        Havoc (0C21000h)  
00C2102D B9 35 12 00 00       mov         ecx,1235h  
00C21032 8B D8                mov         ebx,eax  
00C21034 E8 C7 FF FF FF       call        Havoc (0C21000h)  
00C21039 8B F8                mov         edi,eax  
00C2103B C4 E2 41 F7 F3       shlx        esi,ebx,edi  
00C21040 8B CE                mov         ecx,esi  
00C21042 E8 C9 FF FF FF       call        Sink (0C21010h)  
00C21047 8B CE                mov         ecx,esi  
00C21049 E8 C2 FF FF FF       call        Sink (0C21010h)  
00C2104E 83 E7 01             and         edi,1  
00C21051 C4 E2 41 F7 CB       shlx        ecx,ebx,edi  
00C21056 E8 B5 FF FF FF       call        Sink (0C21010h)  
```

**Comment by GSPP** [comment]
OK, asking a C compiler was a smart idea. x86 is more complicated than I knew. I did not know some of those rules.

* For fun here is GCC: https://goo.gl/N1VOab (Seems like a worse job than VC).
* LLVM seems to twist the calling convention so that the shift result ends up in a convenient register (edi): https://goo.gl/66bkib

I hope my Havoc function was successful in hiding the undefined behavior that this program causes in C languages. Changing `y` to `1` does not change anything.

**Comment by mikedn** [comment]
> LLVM seems to twist the calling convention so that the shift result ends up in a convenient register 

Not really, that compiler targets Linux and Linux uses a different calling convention. Function arguments are passed in edi, esi, edx, ecx while Windows uses ecx, edx, r8, r9.

**Comment by GSPP** [comment]
Understood.

**Comment by mikedn** [comment]
@cmckinsey This should probably be tagged codegen + optimization as the rest of similar issues.

**Comment by cmckinsey** [comment]
As Mike points out most of the register moves are required due to ABI parameter passing conventions, hard requirements around shift index, and callee-save registers for lifetimes live across calls.

But it does look like there is 1 move in the original that could be eliminated, we might consider generated SHLX (AVX2-only which we don't do much of today) on AVX2 machines, and eliminate the unnecessary bitwise & on machines that have modulo semantics on shift index registers.

@sivarv might like to look at the extra register move.

**Comment by mikedn** [comment]
I'm don't see the extra register move when I use CoreCLR. 

@GSPP Is that assembly code generated by .NET 4.6 instead of CoreCLR? You should use CoreCLR for such experiments because it may have changes that aren't yet present in the desktop version. #1163 and its associated fix might be related to that extra move.

**Comment by GSPP** [comment]
@mikedn yes, this was not CodeCLR. I don't recall whether it was 4.5 or 4.6 because I have both available in VMs for testing. I did not realize there were meaningful differences but now I do.

**Comment by mikedn** [comment]
@cmckinsey I think I have a fix for the unnecessary bitwise and. It needs a bit more testing which is time consuming, if you're interested let me know and I'll try to finish it and send a PR.

**Comment by mikedn** [comment]
After #11594 the generated code is:
```asm
       B934120000           mov      ecx, 0x1234
       E8DFFBFFFF           call     ProgramO:Havoc(int):int
       8BF0                 mov      esi, eax
       B935120000           mov      ecx, 0x1235
       E8D3FBFFFF           call     ProgramO:Havoc(int):int
       8BF8                 mov      edi, eax
       8BDF                 mov      ebx, edi
       83E31F               and      ebx, 31
       8BCB                 mov      ecx, ebx
       8BC6                 mov      eax, esi
       D3E0                 shl      eax, cl
       8BC8                 mov      ecx, eax
       E8C7FBFFFF           call     ProgramO:Sink(int)
       8BCB                 mov      ecx, ebx
       8BC6                 mov      eax, esi
       D3E0                 shl      eax, cl
       8BC8                 mov      ecx, eax
       E8BAFBFFFF           call     ProgramO:Sink(int)
       8BCF                 mov      ecx, edi
       83E101               and      ecx, 1
       D3E6                 shl      esi, cl
       8BCE                 mov      ecx, esi
       E8ACFBFFFF           call     ProgramO:Sink(int)
```

An unnecessary `and y, 31` still exists. CSE picked up some of them and the lowering transform added in 11594 can't get rid of them because it sees `x << lcl` rather than `x << (y & 31)`.
